phpBB/phpbb/textformatter/s9e/factory.php

```diff
@@ -0,0 +1,517 @@
+
+		// Set the rendering engine and configure it to save to the cache dir
+		$configurator->rendering->engine = 'PHP';
+		$configurator->rendering->engine->cacheDir = $this->cache_dir;
+		$configurator->rendering->engine->defaultClassPrefix = 's9e_renderer_';
+		$configurator->rendering->engine->enableQuickRenderer = true;
+
+		// Create custom filters for BBCode tokens that are supported in phpBB but not in
+		// s9e\TextFormatter
+		$filter = new RegexpFilter('#^' . get_preg_expression('relative_url') . '$#Du');
+		$configurator->attributeFilters->add('#local_url', $filter);
+		$configurator->attributeFilters->add('#relative_url', $filter);
+
+		// INTTEXT regexp from acp_bbcodes
+		$filter = new RegexpFilter('!^([\p{L}\p{N}\-+,_. ]+)$!Du');
+		$configurator->attributeFilters->add('#inttext', $filter);
+
+		// Create custom filters for Flash restrictions, which use the same values as the image
+		// restrictions but have their own error message
+		$configurator->attributeFilters
+			->add('#flashheight', __NAMESPACE__ . '\\parser::filter_flash_height')
+			->addParameterByName('max_img_height')
+			->addParameterByName('logger');
+
+		$configurator->attributeFilters
+			->add('#flashwidth', __NAMESPACE__ . '\\parser::filter_flash_width')
+			->addParameterByName('max_img_width')
+			->addParameterByName('logger');
+
+		// Create a custom filter for phpBB's per-mode font size limits
+		$configurator->attributeFilters
+			->add('#fontsize', __NAMESPACE__ . '\\parser::filter_font_size')
+			->addParameterByName('max_font_size')
+			->addParameterByName('logger')
+			->markAsSafeInCSS();
+
+		// Create a custom filter for image URLs
+		$configurator->attributeFilters
+			->add('#imageurl', __NAMESPACE__ . '\\parser::filter_img_url')
+			->addParameterByName('urlConfig')
+			->addParameterByName('logger')
+			->addParameterByName('max_img_height')
+			->addParameterByName('max_img_width')
+			->markAsSafeAsURL();
+
+		// Add default BBCodes
+		foreach ($this->get_default_bbcodes($configurator) as $bbcode)
+		{
+			$configurator->BBCodes->addCustom($bbcode['usage'], $bbcode['template']);
+		}
+
+		// Modify the template to disable images/flash depending on user's settings
+		foreach (array('FLASH', 'IMG') as $name)
+		{
+			$tag = $configurator->tags[$name];
+			$tag->template = '<xsl:choose><xsl:when test="$S_VIEW' . $name . '">' . $tag->template . '</xsl:when><xsl:otherwise><xsl:apply-templates/></xsl:otherwise></xsl:choose>';
+		}
+
+		// Load custom BBCodes
+		foreach ($this->dal->get_bbcodes() as $row)
+		{
+			// Insert the board's URL before {LOCAL_URL} tokens
+			$tpl = preg_replace_callback(
+				'#\\{LOCAL_URL\\d*\\}#',
+				function ($m)
+				{
+					return generate_board_url() . '/' . $m[0];
+				},
+				$row['bbcode_tpl']
+			);
+
+			try
+			{
+				$configurator->BBCodes->addCustom($row['bbcode_match'], $tpl);
+			}
+			catch (\Exception $e)
+			{
+				/**
+				* @todo log an error?
```

Comment:
On that note, I've just enabled the creation of unsafe BBCodes to match 3.1's behaviour but at some point in the future the ACP module should be modified to leverage the library's ability to detect unsafe BBCodes. 34add55b6acb48362eb11fb63e07f6eb0a528066